File: algorithms/08_r/graphs/kahn_topological_sort.R
kahn_topological_sort <- function(n, edges) {
  graph <- vector("list", n)
  indegree <- integer(n)
  for (i in 1:n) graph[[i]] <- integer(0)
  for (k in seq_len(nrow(edges))) {
    u <- edges$from[k]
    v <- edges$to[k]
    graph[[u]] <- c(graph[[u]], v)
    indegree[v] <- indegree[v] + 1
  }
  queue <- which(indegree == 0)
  order <- integer(0)
  while (length(queue) > 0) {
    u <- queue[1]
    queue <- queue[-1]
    order <- c(order, u)
    for (v in graph[[u]]) {
      indegree[v] <- indegree[v] - 1
      if (indegree[v] == 0) queue <- c(queue, v)
    }
  }
  if (length(order) == n) order else integer(0)
}

# Example
edges <- data.frame(
  from = c(6, 6, 5, 5, 3, 4),
  to   = c(1, 3, 1, 2, 4, 2)
)
order <- kahn_topological_sort(6, edges)
stopifnot(length(order) == 6)
pos <- integer(6)
pos[order] <- seq_len(6)
stopifnot(all(pos[edges$from] < pos[edges$to]))
print("[R kahn_topological_sort] ... verified")

File: algorithms/05_csharp/graphs/kahn_topological_sort.cs
using System;
using System.Collections.Generic;

namespace Algorithms.Graphs
{
    public static class KahnTopologicalSort
    {
        public static List<int> TopologicalOrder(int n, List<(int from, int to)> edges)
        {
            var graph = new List<int>[n];
            for (int i = 0; i < n; i++) graph[i] = new List<int>();
            var indegree = new int[n];
            foreach (var (from, to) in edges)
            {
                graph[from].Add(to);
                indegree[to]++;
            }

            var queue = new Queue<int>();
            for (int i = 0; i < n; i++)
                if (indegree[i] == 0) queue.Enqueue(i);

            var order = new List<int>();
            while (queue.Count > 0)
            {
                int u = queue.Dequeue();
                order.Add(u);
                foreach (int v in graph[u])
                {
                    if (--indegree[v] == 0) queue.Enqueue(v);
                }
            }

            return order.Count == n ? order : new List<int>();
        }

        public static void Main()
        {
            var edges = new List<(int, int)> { (5, 0), (5, 2), (4, 0), (4, 1), (2, 3), (3, 1) };
            var order = TopologicalOrder(6, edges);
            var pos = new Dictionary<int, int>();
            for (int i = 0; i < order.Count; i++) pos[order[i]] = i;
            bool ok = order.Count == 6;
            foreach (var (u, v) in edges) ok &= pos[u] < pos[v];
            Console.WriteLine(ok ? "[C# KahnTopologicalSort] ... verified" : "[C# KahnTopologicalSort] ... FAILED");
        }
    }
}
